nav2_waypoint_follower/src/waypoint_follower.cpp

```diff
@@ -304,4 +379,58 @@ WaypointFollower::goalResponseCallback(
   }
 }
 
+std::vector<geometry_msgs::msg::PoseStamped>
+WaypointFollower::convertGPSPoses2MapPoses(
+  const std::vector<nav2_msgs::msg::OrientedNavSatFix> & gps_poses,
+  const rclcpp_lifecycle::LifecycleNode::SharedPtr & parent_node,
+  const std::unique_ptr<nav2_util::ServiceClient<robot_localization::srv::FromLL>> & fromll_client)
+{
+  RCLCPP_INFO(parent_node->get_logger(), "Converting GPS waypoints to Map Frame..");
+  rclcpp::Duration transform_tolerance(0, 500);
+
+  std::vector<geometry_msgs::msg::PoseStamped> poses_in_map_frame_vector;
+  for (auto && curr_gps_pose : gps_poses) {
+    auto request = std::make_shared<robot_localization::srv::FromLL::Request>();
+    auto response = std::make_shared<robot_localization::srv::FromLL::Response>();
+    request->ll_point.latitude = curr_gps_pose.position.latitude;
+    request->ll_point.longitude = curr_gps_pose.position.longitude;
+    request->ll_point.altitude = curr_gps_pose.position.altitude;
+
+    fromll_client->wait_for_service((std::chrono::seconds(1)));
+    auto is_conversion_succeeded = fromll_client->invoke(
+      request,
+      response);
+    if (!is_conversion_succeeded) {
+      RCLCPP_ERROR(
+        parent_node->get_logger(),
+        "fromLL service of robot_localization could not convert %i th GPS waypoint to"
+        "Map frame, going to skip this point!"
+        "Make sure you have run navsat_transform_node of robot_localization");
+      continue;
+    } else {
+      // this poses are assumed to be on global frame (map)
+      geometry_msgs::msg::PoseStamped curr_pose_in_map_frame;
+      curr_pose_in_map_frame.header.frame_id = "map";
+      curr_pose_in_map_frame.header.stamp = parent_node->now();
```

Comment:
Better safe starting a timestamp at the start of this function and then using that timestamp for all waypoints. That way they're all the same and you call time `now()` once.